src/logger.ts

```diff
@@ -40,42 +32,80 @@ function getTimestamp(): string {
-    private fd = 0;
+    private fd: number | undefined;
+    private filePath: string | undefined;
     private loggers: Map<string, ILogger> = new Map();
-    private level!: LogLevel;
+    private level!: vscode.LogLevel;
     private logToConsole!: boolean;
     private outputChannel?: vscode.LogOutputChannel;
 
     /**
      * Setup logging for the extension. Logs are dropped unless one or more of
      * `filePath`, `logToConsole`, or `outputChannel` is given.
      *
-     * @param level Only log messages at or above this level, or never if set to `LogLevel.none`.
+     * @param level Only log messages at or above this level, or never if set to `vscode.LogLevel.Off`.
      * @param filePath Write messages to this file.
      * @param logToConsole Write messages to the `console` (Hint: run the "Developer: Toggle Developer Tools" vscode command to see the console).
      * @param outputChannel Write messages to this vscode output channel.
      */
-    public init(level: LogLevel, filePath: string, logToConsole = false, outputChannel?: vscode.LogOutputChannel) {
+    public init(
+        level: vscode.LogLevel,
+        filePath: string,
+        logToConsole = false,
+        outputChannel?: vscode.LogOutputChannel,
+    ) {
         this.level = level;
         this.logToConsole = logToConsole;
         this.outputChannel = outputChannel;
-        if (filePath && level !== LogLevel.none) {
-            try {
-                this.fd = fs.openSync(filePath, "w");
-                this.disposables.push({
-                    dispose: () => fs.closeSync(this.fd),
-                });
-            } catch {
-                // ignore
-            }
-        }
+        this.filePath = filePath;
+
+        this.listenForLogLevelChanges();
+        this.setupLogFile();
     }
 
     public dispose(): void {
         disposeAll(this.disposables);
     }
 
-    private log(level: LogLevel, scope: string, logToOutputChannel: boolean, args: any[]): void {
+    private listenForLogLevelChanges() {
+        this.outputChannel?.onDidChangeLogLevel(
+            (level) => {
+                this.level = level;
+                this.setupLogFile();
+            },
+            undefined,
+            this.disposables,
+        );
+    }
+
+    private setupLogFile() {
+        if (!this.filePath) {
+            return;
+        } else if (this.level !== vscode.LogLevel.Off && this.fd) {
+            return;
+        } else if (this.level === vscode.LogLevel.Off && this.fd) {
+            fs.closeSync(this.fd);
+            this.fd = undefined;
+            return;
+        }
+
+        try {
+            this.fd = fs.openSync(this.filePath, "w");
+            this.disposables.push({
+                dispose: () => {
+                    if (!this.fd) {
+                        return;
+                    }
+
+                    fs.closeSync(this.fd);
+                },
+            });
+        } catch {
+            // ignore
```

Comment:
can we log this somehow? (console.log in the worst case)

Reply:
Yeah for sure. Honestly I don't know why I didn't think to, I was just focused on moving this around. No reason to leave it that way